python/ql/src/experimental/semmle/python/frameworks/Django.qll

```diff
@@ -7,4 +7,313 @@ private import experimental.dataflow.DataFlow
+            exists(DataFlow::StepSummary summary |
+              objects_attr_first_join(t2, attr_name, result, summary) and
+              t = t2.append(summary)
+            )
+          )
+        }
+
+        pragma[nomagic]
+        private predicate objects_attr_first_join(
+          DataFlow::TypeTracker t2, string attr_name, DataFlow::Node res,
+          DataFlow::StepSummary summary
+        ) {
+          DataFlow::StepSummary::step(objects_attr(t2, attr_name), res, summary)
+        }
+
+        /**
+         * Gets a reference to the attribute `attr_name` of an `objects` object.
+         * WARNING: Only holds for a few predefined attributes.
+         */
+        DataFlow::Node objects_attr(string attr_name) {
+          result = objects_attr(DataFlow::TypeTracker::end(), attr_name)
+        }
+
+        /** Gets a reference to the `django.db.models.expressions` object. */
+        private DataFlow::Node expressions(DataFlow::TypeTracker t) {
+          t.start() and
+          result = DataFlow::importNode("django.db.models.expressions")
+          or
+          t.startInAttr("expressions") and
+          result = models()
+          or
+          exists(DataFlow::TypeTracker t2 | result = expressions(t2).track(t2, t))
+        }
+
+        /** Gets a reference to the `django.db.models.expressions` object. */
+        DataFlow::Node expressions() { result = expressions(DataFlow::TypeTracker::end()) }
+
+        /** Gets a reference to the `django.db.models.expressions.RawSQL` class. */
+        private DataFlow::Node classRawSQL(DataFlow::TypeTracker t) {
+          t.start() and
+          result = DataFlow::importNode("django.db.models.expressions.RawSQL")
+          or
+          t.startInAttr("RawSQL") and
+          result = expressions()
+          or
+          exists(DataFlow::TypeTracker t2 | result = classRawSQL(t2).track(t2, t))
+        }
+
+        /** Gets a reference to the `django.db.models.expressions.RawSQL` class. */
+        DataFlow::Node classRawSQL() { result = classRawSQL(DataFlow::TypeTracker::end()) }
+      }
+    }
+  }
+
+  /** A call to the `django.db.connection.cursor.execute` function. */
+  private class DbConnectionExecute extends SqlExecution::Range, DataFlow::CfgNode {
+    override CallNode node;
+
+    DbConnectionExecute() { node.getFunction() = django::db::execute().asCfgNode() }
+
+    override DataFlow::Node getSql() {
+      result.asCfgNode() in [node.getArg(0), node.getArgByName("sql")]
+    }
+  }
+
+  /** A call to the `annotate` function on a model using a `RawSQL` argument. */
+  private class ObjectsAnnotate extends SqlExecution::Range, DataFlow::CfgNode {
+    override CallNode node;
+    CallNode raw;
+
+    ObjectsAnnotate() {
+      node.getFunction() = django::db::models::objects_attr("annotate").asCfgNode() and
+      raw in [node.getArg(0), node.getArgByName(_)] and
+      raw.getFunction() = django::db::models::classRawSQL().asCfgNode()
```

Comment:
Does it need to be a requirement that RawSQL is called in-place as an argument to annotate?

Meaning that the following code is invalid

```py
raw = RawSQL(foo)
Users.objects.annotate(val=raw)
```

---

I'm guessing the answer is no, but I would propose we wait with handling this case until later on. If you could add a test-case showing this unhandled pattern, that would be great :+1:

Reply:
I was hoping for that to be handled, I will certainly add the test-case.